### PR TITLE
Creative: Inventory design overhaul

### PR DESCRIPTION
**Pull request by kilbith**
:gift: 

creative_inv

### CODE REVIEW

**Comment by kilbith** [comment]
Please note I'll not address any suggestion here due to limited spare time and other priorities.

**Comment by NathanSalapat** [comment]
I like it. I'd use a reaction icon, but those aren't available on the mobile site.

**Comment by paramat** [comment]
Thanks, i feel neutral on this.

**Comment by farooghkz** [comment]
@kilbith:What is its usage?

**Comment by DS-Minetest** [comment]
This adds buttons instead of an detached inventory. :confused:

**Comment by farooghkz** [comment]
@DS-Minetest, He did a good job and it seems nice, but I don't know what will be improved with this.

**Comment by octacian** [comment]
:+1: This is much nicer! Great job.

I recently implemented such a setup in my own subgames, Eden and OCD2, and I've found that the experience is much more enjoyable that that of MTG, as you are able to see far more of the important creative inventory items. Plus, you can always click the the crafting tab while holding an item.

File: mods/creative/inventory.lua [hunk 1]
```diff
@@ -8,60 +8,33 @@ function creative.init_creative_inventory(player)
...
+		creative.init_creative_inventory(minetest.get_player_by_name(player_name))
 
 	for name, def in pairs(tab_content) do
 		if not (def.groups.not_in_creative_inventory == 1) and
-				def.description and def.description ~= "" and
-				(def.name:find(inv.filter, 1, true) or
-					def.description:lower():find(inv.filter, 1, true)) then
+		   def.description and def.description ~= ""	   and
+		  (def.name:find(inv.filter, 1, true)		   or
+		   def.description:lower():find(inv.filter, 1, true)) then
```
Comment (SmallJoker) [inline mods/creative/inventory.lua]: Two tabs for additional `if` lines, I think you may use spaces afterwards to line up.

**Comment by SmallJoker** [comment]
Nice feature. Saves space but still allows some item managing in the first eight slots.

**Review by SmallJoker — APPROVED** [review]

**Comment by DS-Minetest** [comment]
:-1: 
Tested, personally I don't like this.
- The buttons instead of the detached inventory are somehow uncomfortably. And also I want the items to rotate, with these buttons impossible.
- When using the buttons in folded out mode, you can only fill those 8 slots, not more.
- On first opening of the inventory in a game session, you have the folded out mode. Maybe it should be vice versa since normally the situation that was before is set to default.
- Getting full stacks might look pretty nice at first but you will never loose one of the stack's items => if you pick up an item of the same type, you will have a new stack which takes another slot. If you want to have a big stack, you can use giveme.

**Comment by GreenXenith** [comment]
> I don't see a big any advantage in seeing more items on one page.

Less pages to sort through.

**Comment by octacian** [comment]
Didn't notice the buttons instead of typical inventory items. Personally not a fan of them either. With a normal inventory, I'd be really happy to see this PR merged.

**Comment by NathanSalapat** [comment]
Shouldn't it be possible to do this with inventory items too. Aren't there just two formspecs that are changed between when clicking on the little toggle icon?
I realize it's a little more complicated than that due to the pages and stuff, but at a core level shouldn't it just be changing between two layouts for the inventory displays.

**Comment by GreenXenith** [comment]
I like the idea but the buttons are :-1:

**Comment by kilbith** [comment]
Item buttons
---------------
Pros:

1. Resizable and easily customizable
2. Less user interaction
3. Mobile devices friendly (consequence of 2.)
4. Less callbacks involved, less complex under the hood
5. Faster to generate than an inventory (consequence of 4.)

Cons:

- No item animation
- Items are always added in a pre-determined slot of the player inventory

**Comment by paramat** [comment]
So i assume the items are buttons that add a full stack of that item to the next open slot of inventory?

> And also I want the items to rotate, with these buttons impossible.

> Getting full stacks might look pretty nice at first but you will never loose one of the stack's items => if you pick up an item of the same type, you will have a new stack which takes another slot.

These points concern me, personally i'm not bothered about rotation but losing it may be unpopular.

**Comment by kilbith** [comment]
~~There's only one simple [condition](https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L3351) to add to prevent a new stack being added to the inventory when an item (that you already own) is picked up. No big deal, however I won't do it.~~

**EDIT:** Actually I checked this in-game and this concern doesn't happen.

**Comment by paramat** [comment]
Ok good. I do like how a stack is added on a button push, when i am testing crafting recipes or crafting something not in creative inventry dragging multiple items out is a pain.

**Comment by octacian** [comment]
:-1: Unless the buttons are replaced with a normal inventory, this is just no. I would be incredibly unhappy to lose item rotation, and when it comes to getting items to test crafting recipes, you can just shift-click to quickly grab several items. And using the proper inventory, we can in the future add improvements to make getting a mass amount of items easier.

**Comment by kilbith** [comment]
Well, you are only a small minority to dislike it. And the most vocal one.

**Comment by ExeterDad** [comment]
Okay, I'll step forward. I much prefer slots over buttons. But I like everything else about it.

**Comment by GreenXenith** [comment]
Perhaps make it an option to use either buttons or slots.

**Comment by octacian** [comment]
Honestly, the buttons instead of a proper inventory are just a workaround for an issue with inventories. What really needs to happen (rather than using the workaround), is the real problems with inventories themselves need to be fixed. In the meantime, I think that things like this should use the technically proper approach in order to make less work after the issues are fixed.

**Comment by Napiophelios** [comment]
>On first opening of the inventory in a game session, you have the folded out mode. Maybe it should be vice versa since normally the situation that was before is set to default.

This is the only objection I have, I too think it should be vice versa.
I think the rest is pretty cool.

**Comment by DS-Minetest** [comment]
The problem that you can now only get one item once out of creative inventory could maybe be fixed by making into the detached inventory full stacks. For one item you would have to right-click instead of left-click, that would be no big change. The only problem, I see with that would be that it could look bad with the numbers.

**Comment by octacian** [comment]
There are better solutions that could be implemented in the future than to make all the stacks full stacks. In the meantime, we'd best use an inventory here in order to make it easier when someone is able to implement such improvements in the engine.

**Comment by paramat** [comment]
Please no slot / button option, let's just choose what's best.
I feel the lack of item rotation will be controversial. Let's wait for more dev input on this.

**Comment by kilbith** [comment]
http://irc.minetest.ru/minetest-dev/2017-07-16#i_5011171

**Pull request closed without merging**